app/apollo/resolvers/cluster.js

```diff
@@ -672,14 +748,21 @@ const clusterResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, cluster_id }, `${queryName} validating`);
+        const cluster = await models.Cluster.findOne({
+          org_id,
+          cluster_id
+        }).lean({ virtuals: true });
+        if(!cluster){
+          throw new NotFoundError(context.req.t('Could not find the cluster with Id {{clusterId}}.', {'clusterId':cluster_id}), context);
+        }
 
-        await validAuth(me, org_id, ACTIONS.UPDATE, TYPES.CLUSTER, queryName, context);
+        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
         validateString( 'cluster_id', cluster_id );
 
-        logger.info({ req_id, user, org_id, cluster_id }, `${queryName} saving`);
+        await validAuth(me, org_id, ACTIONS.UPDATE, TYPES.CLUSTER, queryName, context, [cluster_id, cluster.registration.name || cluster.name]);
+        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating - authorized`);
```

Comment:
- The `validating` log message should be at the beginning of the validation section (i.e. right after the `try {` line).
- Simple checks for valid input should be first
- While the DB query can be before the `validAuth` call, the throwing of NotFoundError should be _after_, to avoid information leakage.
- Similarly it shouldn't return "more than one" error until after it has determined if the user is authorized or not -- if the user isn't authorized on what they are querying, we shouldn't tell them there exist multiple.
- The validAuth call will need to change it's `[ id, name ]` argument depending on whether the cluster was found in the DB or not.